aae_progress_report and other functions needed for Tictac AAE CLI treestatus subcommand

hmmr opened the pull request:
This PR adds ~getters for some controller state records~ functions needed for [TicTac AAE and NextGenRepl CLI commands](https://github.com/orgs/OpenRiak/discussions/10) (main body of work is in this [PR in riak_kv](https://github.com/OpenRiak/riak_kv/pull/55)):

* `aae_controller:aae_produce_progress_report/1`, to collect and generate a "progress report" of tree (re)building status, used by `riak admin tictacaae treestatus` command;
* `aae_controller:aae_schedulenextrebuild/2`, which can be used to trigger a tree rebuild, at a specified delay.

ThomasArts commented on src/aae_controller.erl:
```diff
@@ -24,34 +24,41 @@
             hash_clocks/2,
             wrapped_splitobjfun/1]).
 
+-export([get_key_store/1,
```
Since you are not supposed to make the state of a `gen_server` visible outside the `gen_server` itself, one should not export functions that take that state as input argument.
This is possibly a remainder of an experiment, but I propose to drop those functions and in particular not exporting them.

hmmr replied:
Yeah, it's wip at the moment. I should probably have taken it private until it's ready.

Commit:
touch object_splitfun directly for purposes of get/setting storeheads

this, to avoid introducing dependency on riak_kv, where
riak_object:aae_from_object_binary/1 lives, which actually creates the
object_splitfun function.

martinsumner commented:
[The acceptance criteria](https://github.com/OpenRiak/kv_index_tictactree#contributing-and-testing) requires 100% code coverage, so some tests will be required.

I don't think they have to be too involved in the majority cases (just confirm that we can set, then get what we expect).  The more difficult one may be the dynamic setting of the object split fun.  I assume that this is expected to require a prompted rebuild before it can be used, so perhaps a ct test of that process would be useful (i.e. have one function, test query, update function, test query with no change in output, update some objects, test query to show partial change, force rebuild, re-test query).

hmmr commented:
> [The acceptance criteria](https://github.com/OpenRiak/kv_index_tictactree#contributing-and-testing) requires 100% code coverage, so some tests will be required.

There is an integration test covering the whole lot, in https://github.com/OpenRiak/riak_test/pull/38. I will provide basic eunit tests, for completeness' sake.

> The more difficult one may be the dynamic setting of the object split fun. I assume that this is expected to require a prompted rebuild before it can be used, so perhaps a ct test of that process would be useful (i.e. have one function, test query, update function, test query with no change in output, update some objects, test query to show partial change, force rebuild, re-test query).

I'll think about this, too. However, I don't think it is critical to achieve the immediate effect with this one; just, users will be made aware that setting `storeheads` will only take effect on subsequent tree rebuilds.

Commit:
aae_schedulenextrebuild/2

Commit:
get/setters of various state fields needed for riak admin tictacaae

Commit:
ct test for get/set_rebuild_schedule

Commit:
ct for get/set_storeheads

Commit:
uncomment, unbreak other tests in basic_SUITE

Commit:
remove extra, redundant steps in get_set_storeheads

Commit:
avoid exposing internal pids, do produce_aae_progress_report on site

Commit:
more expressive function names for aae progress report work

martinsumner commented:
I think that the function `aae_schedulenextrebuild/2` might be better named `aae_prompt_nextrebuild/2`.  It needs to be clearer on the difference with `aae_set_rebuild_schedule/2`.  With `aae_schedulenextrebuild/2` the schedule is not changed (i.e. the recurring frequency of rebuilds), it is just that the next rebuild is directly prompted to occur after delay.

martinsumner commented:
It would be good to add a test which uses `aae_schedulenextrebuild/2`, and confirms that the next rebuild time is indeed changed by looking at the output of `aae_produce_progress_report/1`.

martinsumner commented:
I've been trying for both leveled and kv_index_tictactree to avoid passing the state record to functions outside of the standard gen_server callbacks (handle_call/cast etc).  So ideally produce_report/1 would either be contained within the handle_call, or be a 3-arity function that receives the three components of state it requires.

This unwritten guideline is already broken elsewhere, so it is not a critical issue if it is also broken here.

Commit:
aae_schedulenextrebuild/2 better named aae_prompt_nextrebuild

Commit:
avoid passing entire state record for funs that only need some fields

Commit:
prompt_nextrebuild done right, with a test

martinsumner commented:
There are some coverage misses (in the `aae_controller` module) to be resolved:

```
242 |   | -spec aae_get_object_splitfun(pid()) -> function().
-- | -- | --
243 |   | %% @doc
244 |   | %% Get object_splitfun field. It is only used to infer the value of 'storeheads'.
245 |   | aae_get_object_splitfun(Pid) ->
246 | :-( | gen_server:call(Pid, get_object_splitfun, ?SYNC_TIMEOUT).
```

Is the function actually going to be used i.e. how is the value to be inferred?  If we know how it is to be used/inferred, then there should be a test that shows that it works.

```
produce_report(KeyStore, NextRebuild, TreeCaches) ->
--
1220 | 1 | KeyStoreCurrentStatus =
1221 |   | if is_pid(KeyStore) ->
1222 | 1 | element(1, aae_keystore:store_currentstatus(KeyStore));
1223 |   | el/=se ->
1224 | :-( | not_running
1225 |   | end,
```

I think the aae_keystore must always be a PID, and so the else condition shouldn't exist.  As part of this the state record definition should be changed to remove the `undefined` case.  I suspect this is a legacy of some old test (and dialyzer doesn't seem to complain if `undefined` is removed).

Commit:
callers of produce_report should assume key_store is always a good pid

hmmr commented:
> There are some coverage misses (in the `aae_controller` module) to be resolved:
> 
> ```
> 242 |   | -spec aae_get_object_splitfun(pid()) -> function().
> -- | -- | --
> 243 |   | %% @doc
> 244 |   | %% Get object_splitfun field. It is only used to infer the value of 'storeheads'.
> 245 |   | aae_get_object_splitfun(Pid) ->
> 246 | :-( | gen_server:call(Pid, get_object_splitfun, ?SYNC_TIMEOUT).
> ```
> 
> Is the function actually going to be used i.e. how is the value to be inferred? If we know how it is to be used/inferred, then there should be a test that shows that it works.

This is how the value of storeheads is "inferred": https://github.com/TI-Tokyo/riak_kv/blob/tiot/openriak-3.4/tictacaae-and-nextgenrepl-cli-commands-clique/src/riak_kv_vnode.erl#L1381-L1384. It is compared to a function object that is [created](https://github.com/TI-Tokyo/riak_kv/blob/tiot/openriak-3.4/tictacaae-and-nextgenrepl-cli-commands-clique/src/riak_kv_vnode.erl#L355) and used when `aae_controller:aae_start/8` is called with `StoreHeads = true`.

This storeheads flag is problematic in that in real world, the splitfunction variants corresponding to the flag's value are constructed using `riak_object:aae_from_object_binary`, which comes from riak_kv -- which cannot be made a dependency of kv_index_tictactree because it itself is a dependency of riak_kv.

Comparing functions is borderline acceptable, but it works, and bar a new, dedicated `storeheads` flag in aae_controller state that would be set alongside `object_splitfun` and only consulted by `get_storeheads`, I am struggling to see a good solution.

martinsumner commented:
OK, fetching the fun via the API is OK in that case.  Just add a test to set the function, get it and compare it.

With this change, the PR can be approved and merged.

Commit:
splitfun get/set ct test
